provers/groth16/src/setup.rs

```diff
@@ -0,0 +1,131 @@
+    // [K_{k+1}(τ)]_1, [K_{k+2}(τ)]_1, ..., [K_n(τ)]_1
+    // where K_i(τ) = ƍ^{-1} * (β*l(τ) + α*r(τ) + o(τ))
+    // and "k" is the number of public inputs
+    pub prover_k_tau_g1: Vec<G1Point>,
+    // [delta^{-1} * t(τ) * tau^0]_1, [delta^{-1} * t(τ) * τ^1]_1, ..., [delta^{-1} * t(τ) * τ^m]_1
+    pub z_powers_of_tau_g1: Vec<G1Point>,
+}
+
+struct ToxicWaste {
+    tau: FrElement,
+    alpha: FrElement,
+    beta: FrElement,
+    gamma: FrElement,
+    delta: FrElement,
+}
+
+impl ToxicWaste {
+    pub fn new() -> Self {
+        Self {
+            tau: sample_fr_elem(),
+            alpha: sample_fr_elem(),
+            beta: sample_fr_elem(),
+            gamma: sample_fr_elem(),
+            delta: sample_fr_elem(),
+        }
+    }
+}
+
+pub fn setup(qap: &QAP) -> (ProvingKey, VerifyingKey) {
+    let g1: G1Point = Curve::generator();
+    let g2: G2Point = TwistedCurve::generator();
+
+    let tw = ToxicWaste::new();
+
+    let num_of_total_inputs = qap.l.len();
+
+    // [A_i(τ)]_1, [B_i(τ)]_1, [B_i(τ)]_2, [K_i(τ)]_1
+    let mut l_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut r_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut r_tau_g2: Vec<G2Point> = Vec::with_capacity(num_of_total_inputs);
+    let mut verifier_k_tau_g1: Vec<G1Point> = Vec::with_capacity(qap.num_of_public_inputs);
+    let num_of_private_inputs = num_of_total_inputs - qap.num_of_public_inputs;
+    let mut prover_k_tau_g1: Vec<G1Point> = Vec::with_capacity(num_of_private_inputs);
+
+    let delta_inv = tw.delta.inv().unwrap();
+    let gamma_inv = tw.gamma.inv().unwrap();
+    for i in 0..num_of_total_inputs {
+        let l_i_tau = qap.l[i].evaluate(&tw.tau);
+        let r_i_tau = qap.r[i].evaluate(&tw.tau);
+
+        l_tau_g1.push(g1.operate_with_self(l_i_tau.representative()));
+        r_tau_g1.push(g1.operate_with_self(r_i_tau.representative()));
+        r_tau_g2.push(g2.operate_with_self(r_i_tau.representative()));
+
+        let o_i_tau = qap.o[i].evaluate(&tw.tau);
+        let k_i_tau_unshifted = &tw.beta * &l_i_tau + &tw.alpha * &r_i_tau + &o_i_tau;
+        if i < qap.num_of_public_inputs {
+            // Public variables
+            let k_i_tau = &gamma_inv * k_i_tau_unshifted;
+            verifier_k_tau_g1.push(g1.operate_with_self(k_i_tau.representative()));
+        } else {
+            // Private variables
+            let k_i_tau = &delta_inv * k_i_tau_unshifted;
+            prover_k_tau_g1.push(g1.operate_with_self(k_i_tau.representative()));
+        }
+    }
+
+    // [delta^{-1} * t(τ) * τ^0]_1, [delta^{-1} * t(τ) * τ^1]_1, ..., [delta^{-1} * t(τ) * τ^m]_1
+    let t_tau_times_delta_inv = &delta_inv * qap.vanishing_polynomial().evaluate(&tw.tau);
+    let z_powers_of_tau_g1: Vec<G1Point> = (0..qap.num_of_gates() + 1)
+        .map(|exp: usize| {
+            g1.operate_with_self(
+                (&t_tau_times_delta_inv * tw.tau.pow(exp as u128)).representative(),
```

Comment:
Here I'd take advantage of the fact that tau powers can be obtained using a successors

Reply:
done